clients/go/admin/auth_interceptor.go

```diff
@@ -0,0 +1,74 @@
+package admin
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/flyteorg/flyteidl/clients/go/admin/cache"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"github.com/flyteorg/flytestdlib/logger"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"google.golang.org/grpc"
+)
+
+// ResetTokenSource will attempt to build a TokenSource given the anonymously available information exposed by the server.
+// Once established, it'll invoke CustomHeaderTokenSource.Reset() on tokenSource to populate it with the appropriate values.
+func ResetTokenSource(ctx context.Context, cfg *Config, tokenCache cache.TokenCache, tokenSource *CustomHeaderTokenSource) error {
+	authMetadataClient, err := InitializeAuthMetadataClient(ctx, cfg)
+	if err != nil {
+		// EngHabu: I don't know why did we use to panic here. Generally speaking, libraries should error on the side of
+		// 	returning errors instead.
+		return fmt.Errorf("failed to initialize Auth Metadata Client. Error: %w", err)
+	}
+
+	tokenSourceProvider, err := NewTokenSourceProvider(ctx, cfg, tokenCache, authMetadataClient)
+	if err != nil {
+		return fmt.Errorf("failed to initialize token source provider. Err: %w", err)
+	}
+
+	clientMetadata, err := authMetadataClient.GetPublicClientConfig(ctx, &service.PublicClientAuthConfigRequest{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch client metadata. Error: %v", err)
+	}
+
+	tSource, err := tokenSourceProvider.GetTokenSource(ctx)
+	if err != nil {
+		return err
+	}
+
+	tokenSource.Reset(tSource, clientMetadata.AuthorizationMetadataKey, cfg.UseInsecureConnection)
```

Comment:
we are using tokenSource and tSource which is confusing to read. May be can tokenSource be renamed to singletonTokenSource

Reply:
plz let me know what do you think now